Add subscription_start to registration receipt

sr-gi opened the pull request:
Fixes #76 

## Changes:

### Common
- Updates the protos to add an additional field `RegisterRequest`. Notice this is a breaking change. Previous protos won't work with this one.

### TEOS
- Updates the gRPC server to account for the updated protos
- Updates the database. The `users` table now also stores the `subscription_start` field. This is also a breaking change.
- Adds `subscription_start` to `UserInfo` in the `Gatekeeper`
- Sets constants in tests_utils to reduce the tests boilerplate

### CLN plugin
- Updates the database. `towers` does not store the `subscription_expiry` anymore (it was being stored both there an in the `subscription_receipt`). 
    - The `subscription_start` field is added to `subscription_receipts` and the primary key updated from `tower_id` to `(tower_id, subscription_expiry)`.
    - When info regarding a tower is loaded now, it is done from both the `towers` table and the `subscription_receipts` table, picking the latest receipt for the latter.
    - Adds `load_registartion_receipt(tower_id, user_id)` that loads the latest registration receipt of a given (user_id, tower_id) pair. The latest is determined by the one with the `subscription_expiry` further into the future.
    - `store_tower_record` SQLite query has been updated from `INSERT OR REPLACE` to `UPSERT`. The rationale is that the former does delete + insert on replace, triggering the `ON DELETE` triggers for foreign keys.
- Adds `subscription_start` to both `TowerInfo` and `TowerSummary`

meryacine reviewed (commented):
ACK [`05cd387`](https://github.com/talaia-labs/rust-teos/pull/77/commits/05cd3877b49718d3556b73776e18cf9a26101834)
Looks good. I think we can address the introduced FIXMEs in this PR.

sr-gi commented on watchtower-plugin/src/dbm.rs:
```diff
@@ -207,12 +240,24 @@ impl DBM {
+
+            let mut stmt = self
+            .connection
+            .prepare("SELECT subscription_start, subscription_expiry FROM registration_receipts WHERE tower_id = ?1 AND subscription_expiry = (SELECT MAX(subscription_expiry) FROM registration_receipts WHERE tower_id = ?1)")
```
Umm, this also passes tests:

```
SELECT tw.tower_id, tw.net_addr, tw.available_slots, rr.subscription_start, rr.max_se
FROM towers AS tw
JOIN (SELECT tower_id, subscription_start, MAX(subscription_expiry) AS max_se
      FROM registration_receipts
      GROUP BY tower_id) AS rr
ON (tw.tower_id = rr.tower_id)
```

Meaning the tests are insufficient, or the query can be simplified

sr-gi replied:
Do double check, b/c my SQL is a bit Rusty, but it is indeed passing the current tests for `watchtower-plugin`

meryacine replied:
In [`here`](https://dba.stackexchange.com/questions/141594/how-select-column-does-not-list-in-group-by-clause) and [`here`](https://tableplus.com/blog/2019/10/select-column-not-in-group-by.html), this gives an error. So it's only SQLite being nice and fulfilling the query.
The tests are indeed not covering the case where more than one registration per tower exists.

meryacine replied:
Aside from the test not being hard enough. Your query would actually work even on the hardened test.
I see this is because the grouping function used is `MAX()`. I tried doing a similar thing with `COUNT()` and it failed.
I wouldn't count on it though, cuz we might use another SQL engine later which would reject this query.

meryacine replied:
> The tests are indeed not covering the case where more than one registration per tower exists.
In here I meant the test `test_store_load_towers`, which should test the mentioned case.

meryacine replied:
After hardening the test, nope.
Because the thing about the available slots being overwritten with random numbers.
the [hardened test](https://github.com/meryacine/rust-teos/commit/25e0dbc04336b5dfa38ab8f0d1b3d0fc66bf1d70#diff-3af2742aaf4be3e870d3a4e8fab4d0b8990e271654af4e74b9eacb62db0df72cL696-L719) would fail with the 2 sql queries above and with the current code too.

meryacine commented on watchtower-plugin/src/dbm.rs:
```diff
@@ -139,24 +140,25 @@ impl DBM {
-            ],
+        let tx = self.get_mut_connection().transaction().unwrap();
+        tx.execute(
+            "INSERT INTO towers (tower_id, net_addr, available_slots) VALUES (?1, ?2, ?3) ON CONFLICT (tower_id) DO UPDATE SET net_addr = ?2, available_slots = ?3",
```
Shouldn't the slots be added to the already existing ones?
Realized this while hardening `test_store_load_towers` test: I needed to store only the receipt with the greatest sub expiry but the tower's `available_slots` was messed up because we update it with `store_tower_record` using random receipts.

sr-gi replied:
That may be due to how the test helpers are built. In theory, when receiving a new registration receipt the tower already adds either (or both) the slots and time to the current one.

If we update it with a new random receipt, this won't hold given the receipt is not tied to the previous one in this case.

meryacine commented on watchtower-plugin/src/dbm.rs:
```diff
@@ -139,24 +140,25 @@ impl DBM {
+            params![tower_id.to_vec(), net_addr, receipt.available_slots()],
         )
+        .map_err( Error::Unknown)?;
+        // FIXME: This may need a guard. What if a tower returns the same end_time twice?
```
Yeah we better send a combined receipt from the tower.
If we know that our sub isn't over yet, we should verify the validity of the received receipt and if it's fine we replace the old one by it.
This case we won't actually need to keep a `registration_id`.

I think we can also have `tower_id` only be the PK, because, why would we need to store old receipts?

sr-gi replied:
This holds as long as there is no gap in the subscriptions:
```
|------------------|---------------------|
	S1		    S2
```

If the client has an appointment receipt for `A1` sent to the tower during `S1`, and there is no gap between `S1` and `S2`, then the user cannot trick the tower, having `S2` would suffice.

```
|------------------|  GAP  |---------------------|
	S1			     S2
```

If there is a gap between `S1` and `S2`, if we allow `S2` and `A1` to be enough to prove misbehaving, then the user can wait until its data is wiped from the tower (GAP), register again (`S2`) and trigger `A1`. The tower won't be able to react to `A1`.

Therefore, the misbehaving proof should include registration receipts from `A1` start_block to the block the breach was not penalized (without gaps). If the user does not interrupt the subscription, then the tower can simply batch the receipts and both will only need to store a single one.

LMK if this logic holds or if you find anything weird.

sr-gi commented:
@meryacine added the updated hardened test and addressed the fixmes:

- `WTClient::add_update_tower` now expects both the slots and expiry to be higher than the previous subscription iteration.
- Added a tests for the unguarded `dbm::store_tower_record` to tests that it will fail if used unguarded.

wrt to the first point, we will need to update those contains in the future, but now they do match the ones the tower follows anyway. With this, we at least prevent someone from modifying the tower code and trying to be silly about subscription updates trying to reduce either the time or the slot count (or both).

meryacine reviewed (approved):
LGTM
